src/python/pants/backend/python/dependency_inference/scripts/import_parser.py

```diff
@@ -39,9 +30,17 @@ def __init__(self, package_parts, contents):
         self.strong_imports = {}
         self.weak_imports = {}
         self._weaken_strong_imports = False
+        if os.environ["STRING_IMPORTS"] == "y":
+            # This regex is used to infer imports from strings, e.g .
+            #  `importlib.import_module("example.subdir.Foo")`.
+            self._string_import_regex = re.compile(
+                r"^([a-z_][a-z_\d]*\.){" + os.environ["MIN_DOTS"] + r",}[a-zA-Z_]\w*$", re.UNICODE
+            )
+        else:
+            self._string_import_regex = None
 
     def maybe_add_string_import(self, node, s):
-        if os.environ["STRING_IMPORTS"] == "y" and STRING_IMPORT_REGEX.match(s):
+        if self._string_import_regex and self._string_import_regex.match(s):
```

Comment:
Alternatively we could always compile the regex, but make it `"(?!x)x"` (or whatever your favorite character it) so that we don't have to do the truthy check, we can always try and match :man_shrugging:

Reply:
We're likely talking in the order of sub-ms, maybe a single ms so timing isn't that big of a deal. I think simplicity is the main motivator for this PR so whatever you think is simpler. :smile:

Reply:
Lol ok we're talking _nanoseconds_ here :joy:

Reply:
And for comparison here's the current timing:
```
Benchmark 1: STRING_IMPORTS=N MIN_DOTS=2 python src/python/pants/backend/python/dependency_inference/scripts/import_parser.py src/python/pants/option/global_options.py
  Time (mean ± σ):      32.7 ms ±   2.1 ms    [User: 27.0 ms, System: 5.5 ms]
  Range (min … max):    28.2 ms …  37.3 ms    85 runs
```

and with your change

```
Benchmark 1: STRING_IMPORTS=N MIN_DOTS=2 python src/python/pants/backend/python/dependency_inference/scripts/import_parser.py src/python/pants/option/global_options.py
  Time (mean ± σ):      32.2 ms ±   2.0 ms    [User: 27.4 ms, System: 4.8 ms]
  Range (min … max):    27.8 ms …  36.3 ms    85 runs
````